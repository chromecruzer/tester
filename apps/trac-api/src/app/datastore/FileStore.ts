
import path from "path";
import {AuditRecord, Comparisons, EmailSettings} from "@trac/datatypes";

import util from "util";
import DataStoreUtilities from "./DataStoreUtilities";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class FileStore {
  private json: ((Comparisons | AuditRecord | EmailSettings) | null) = null;
  private filename;
  private dsu: DataStoreUtilities;
  constructor(storagePath, filename) {
    this.filename = path.join(storagePath, `${filename}.json`);
    this.dsu = new DataStoreUtilities();
  }
  public async get() {
    // console.log(`cached content is ${dump(this.json)}`);
    if(this.json) {
      return this.json;
    }
    return this.read();
  }
  private async read() {
    if(!await this.dsu.check(this.filename)) {
      return null;
    }
    return this.dsu.retrieve(this.filename).then(read => {
      // console.log(`retrieved content ${dump(read)}`)
      this.json = read;
      return read;
    });
  }
  public async set(json) {
    this.json = json;
    // console.log(`Writing to ${this.filename} content ${dump(json)}`);
    return this.dsu.store(this.filename, json);
  }
  public async remove() {
    this.json = null;
    return this.dsu.remove(this.filename)
  }
}
