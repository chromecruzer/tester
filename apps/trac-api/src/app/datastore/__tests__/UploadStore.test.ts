import FileStore from "../FileStore";

describe('FileStore', () => {
  let mockDsu, dataStore;
  beforeEach(() => {
    const storagePath = 'some/root/directory';
    mockDsu = {
      mkdir: jest.fn().mockReturnValue(Promise.resolve()),
      check: jest.fn().mockReturnValue(Promise.resolve(true)),
      remove: jest.fn().mockReturnValue(Promise.resolve()),
      store: jest.fn().mockReturnValue(Promise.resolve()),
      retrieve: jest.fn().mockReturnValue(Promise.resolve({}))
    };
    dataStore = new FileStore(storagePath, 'session-token');
    dataStore.dsu = mockDsu;
  });
  describe('get', () => {
    it('should get the stored data', () => {
      const expected = {retrieved: 'json data'};
      mockDsu.retrieve.mockReturnValue(Promise.resolve(expected));
      return dataStore.get().then(actual => {
        expect(actual).toEqual(expected);
      })
    });
    it('should get the cached data', async () => {
      const expected = {retrieved: 'json data'};
      mockDsu.retrieve.mockReturnValue(Promise.resolve(expected));
      await dataStore.get(); // should also store in the cache.
      mockDsu.retrieve.mockReturnValue(Promise.resolve({not_retrieved:'json data'}));
      return dataStore.get().then(actual => {
        expect(actual).toEqual(expected);
      })
    });
  });
  describe('set', () => {
    it('should set both cached and stored', () => {
      const expected = {submitted: 'json data'};
      return dataStore.set(expected).then(() => {
        expect(dataStore.json).toEqual(expected);
        expect(mockDsu.store.mock.calls).toMatchSnapshot();
      })
    });
  });
  describe('remove', () => {
    it('should remove both stored and cached data', async () => {
      await dataStore.set({tobedeleted: 'json data'});
      return dataStore.remove().then(() => {
        expect(dataStore.json).toBeNull();
        expect(mockDsu.remove.mock.calls).toMatchSnapshot();
      })
    });
  });
});
