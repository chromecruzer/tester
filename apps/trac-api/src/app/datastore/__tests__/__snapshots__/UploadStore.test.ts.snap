// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileStore remove should remove both stored and cached data 1`] = `
Array [
  Array [
    "some/root/directory/session-token.json",
  ],
]
`;

exports[`FileStore set should set both cached and stored 1`] = `
Array [
  Array [
    "some/root/directory/session-token.json",
    Object {
      "submitted": "json data",
    },
  ],
]
`;
